pyproject.toml

```diff
@@ -40,3 +41,10 @@ myst-parser = "^3.0.1"
 [build-system]
 requires = ["poetry-core"]
 build-backend = "poetry.core.masonry.api"
+
+[tool.coverage.report]
+exclude_lines = [
+    "pragma: no cover",
+    "if TYPE_CHECKING:",
+    "raise NotImplementedError",
```

Comment:
If the `DomainTransformer` were an `abc.ABC` (which it probably should be), then we wouldn't need to exclude `NotImplementedError` anymore, correct? Or is this a standard line to include?

Reply:
I forgot, I'll fix it soon.

Reply:
I forgot.
https://github.com/bayesian-optimization/BayesianOptimization/blob/d305c1b6584432841a60120cf45e5a2f1b68bef0/bayes_opt/domain_reduction.py#L25
https://github.com/bayesian-optimization/BayesianOptimization/blob/d305c1b6584432841a60120cf45e5a2f1b68bef0/bayes_opt/domain_reduction.py#L29
These two places seem to be areas that are not tested and don't need to be tested.
But if `"raise NotImplementedError",` is removed, it will be reflected in the code coverage.

Should I leave them in, or comment `pragma: no cover`?

Reply:
I think the move here would be to convert the `DomainTransformer` into an `abc.ABC` and use `@abstractmethod` here.  If you want, you can do that in this PR, otherwise we can leave things as-is.

Reply:
> I think the move here would be to convert the `DomainTransformer` into an `abc.ABC` and use `@abstractmethod` here. If you want, you can do that in this PR, otherwise we can leave things as-is.

I'll fix it soon.

Reply:
> > I think the move here would be to convert the `DomainTransformer` into an `abc.ABC` and use `@abstractmethod` here. If you want, you can do that in this PR, otherwise we can leave things as-is.
> 
> I'll fix it soon.

By the way, I have a slight problem with the code coverage (2 lines), is this okay?

![스크린샷 2024-07-17 오후 11 47 18](https://github.com/user-attachments/assets/c1db35b8-6ef3-4b63-a3ae-1929e5cf14de)

Reply:
I would handle this similar to the case of the acquisition function: Considering that the methods now have the @abstractmethod decorator you don't need to raise these Errors anyways. See [here](https://github.com/bayesian-optimization/BayesianOptimization/blob/bc9af23981cb22b98344424b5c2544b741b74938/bayes_opt/acquisition.py#L37) for comparison.

Reply:
> I would handle this similar to the case of the acquisition function: Considering that the methods now have the @AbstractMethod decorator you don't need to raise these Errors anyways. See [here](https://github.com/bayesian-optimization/BayesianOptimization/blob/bc9af23981cb22b98344424b5c2544b741b74938/bayes_opt/acquisition.py#L37) for comparison.

removing it from the code itself, I understand.



> If you don't need a base class with functionality, you can also use [protocols](https://docs.python.org/3/library/typing.html#typing.Protocol) in which case you don't need the NotImplementedError's

What do you think of the suggestion to use `Protocol` instead of `ABC`?

Reply:
I would prefer the ABC solution, because I'd rather not add another pattern to the code. I understand that the situation here is distinct from `acquisition.AcquisitionFunction` (because this one is literally nothing but a template) but I think with how small this codebase is they're not distinct enough to differentiate imo.